scrypto/src/core/data.rs

```diff
@@ -0,0 +1,46 @@
+use crate::buffer::*;
+use crate::core::DataAddress;
+use crate::engine::api::RadixEngineInput::WriteData;
+use crate::engine::call_engine;
+use sbor::rust::cell::RefMut;
+use sbor::rust::ops::{Deref, DerefMut};
+use sbor::Encode;
+
+pub struct DataRef<'a, V: Encode> {
+    pub value: RefMut<'a, V>,
+}
+
+impl<'a, V: Encode> Deref for DataRef<'a, V> {
+    type Target = V;
+
+    fn deref(&self) -> &Self::Target {
+        self.value.deref()
+    }
+}
+
+pub struct DataRefMut<'a, V: Encode> {
+    pub address: DataAddress,
+    pub value: RefMut<'a, V>,
+}
+
+impl<'a, V: Encode> Drop for DataRefMut<'a, V> {
+    fn drop(&mut self) {
+        let bytes = scrypto_encode(self.value.deref());
+        let input = WriteData(self.address.clone(), bytes);
+        let _: () = call_engine(input);
```

Comment:
Do you think it makes more sense to have `DataValue` to flush instead of `RefMut`?

Reply:
`DataRefMut` is used by both `DataValue` and `KeyValueStore`, both of which needs this flushing mechanism.

Reply:
Okay, happy to start as-is.